### PR TITLE
PowerShell Implementation

### PR DESCRIPTION
**Pull request by tohling**
Updated functions support for PowerShell.  Added features include:

1.	Running PowerShell in-process
2.	Running PowerShell custom modules – We will support script (*.psm1), binary(*.dll), and manifest (*.psd1) modules.  This would be the most common 3 out of 4 module types.
3.	Printing output and errors in real-time.
4.	Running Azure PowerShell cmdlets.

### CODE REVIEW

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionDescriptionProvider.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.ObjectModel;
+using Microsoft.Azure.WebJobs.Script.Binding;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class PowerShellFunctionDescriptionProvider : FunctionDescriptorProvider
```
Comment (fabiocav) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionDescriptionProvider.cs]: Should this be called `PowerShellFunctionDescriptorProvider`?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+
+                    _script = GetScript();
+                    powerShellInstance.AddScript(_script, true);
+
+                    PSDataCollection<PSObject> outputCollection = new PSDataCollection<PSObject>();
+                    outputCollection.DataAdded += OutputCollectionDataAdded;
+
+                    powerShellInstance.Streams.Error.DataAdded += ErrorDataAdded;
+
+                    IAsyncResult result = powerShellInstance.BeginInvoke<PSObject, PSObject>(null, outputCollection);
```
Comment (fabiocav) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: We should pass a callback to BeginInvoke so we can:

1. Properly call EndInvoke
2. Eliminate the loop in line 178 and just await for the completion instead of calling sleep

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+            }
+
+            SetProcessEnvironmentVariables(environmentVariables);
+        }
+
+        private void ReloadScript()
+        {
+            TraceWriter.Verbose(string.Format(CultureInfo.InvariantCulture, "Script for function '{0}' changed. Reloading.", Metadata.Name));
+
+            _host.RestartEvent.Set();
```
Comment (fabiocav) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Do we need to restart the host for PS function changes?

**Comment by fabiocav** [comment]
@tohling you may want to rebase since there were some changes that should enable you to get a clean build.

File: src/WebJobs.Script/Description/PowerShell/PowerShellConstants.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        public const char UnderscoreChar = '_';
+        public const char AdditionChar = '+';
+        public const string StackTraceScriptBlock = @"<ScriptBlock>, <No file>";
+
+        // Output constants
+        public const string CategoryInfoLabel = "CategoryInfo          :";
+        public const string FullyQualifiedErrorIdLabel = "FullyQualifiedErrorId :";
+
+        // Environment variables constants
+        public const string ResEnvVarKey = "res";
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellConstants.cs]: Where are these used? I did a search on this PR and I don't see it.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+            IBinderEx binder = (IBinderEx)parameters[2];
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters[3];
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            FunctionStartedEvent startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                object convertedInput = input;
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: It's a TODO on me to refactor some of this common code out so it is shared rather than duplicated across the various providers (PowerShell and Script).

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+                        runtimeException.ErrorRecord.InvocationInfo.PositionMessage));
+                }
+
+                TraceWriter.Verbose(string.Format("PowerShell Function completed (Failure, Id={0})", invocationId));
+                throw;
+            }
+            finally
+            {
+                _metrics.EndEvent(startedEvent);
+                DeleteProcessEnvironmentVariables(_environmentVariables);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Hmmm, this might be a problem. In the ScriptInvoker case, these environment variables are being injected into a child process, so they don't pollute the parent process, and there is not possibility of collisions.

In this case, if you have N instances of a function running concurrently, won't they overwrite each others state? I think we need a way to inject these into the PowerShell host so they aren't parent env vars.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+            }
+
+            // The ScriptHost is already monitoring for changes to function.json, so we skip those
+            string fileExtension = Path.GetExtension(e.Name);
+            if (WatchedFileTypes.Contains(fileExtension))
+            {
+                _reloadScript();
+            }
+        }
+        // private void InvokePowerShellScript()
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: nit: remove comment

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+
+                    IAsyncResult result = powerShellInstance.BeginInvoke<PSObject, PSObject>(null, outputCollection);
+
+                    while (result.IsCompleted == false)
+                    {
+                        System.Threading.Thread.Sleep(50);
+                    }
+
+                    // Clean up and remove module
+                    if (_moduleFiles.Any())
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Why do we have to do these removals? Can't we just let it dispose itself since its IDisposable?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+                    {
+                        // TODO: Handle other content types? (E.g. byte[])
+                        if (request.Content != null && request.Content.Headers.ContentLength > 0)
+                        {
+                            convertedInput = ((HttpRequestMessage)input).Content.ReadAsStringAsync().Result;
+                        }
+                    }
+                }
+
+                TraceWriter.Verbose(string.Format("PowerShell Function started (Id={0})", invocationId));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Since you started this work, I overhauled the default verbosity level - it is now Info, not Verbose. See the other script provider. These should be Info level logs, otherwise they won't show up by default.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+
+                TraceWriter.Verbose(string.Format("PowerShell Function completed (Success, Id={0})", invocationId));
+            }
+            catch (Exception exception)
+            {
+                startedEvent.Success = false;
+                RuntimeException runtimeException = exception as RuntimeException;
+                if (runtimeException != null)
+                {
+                    TraceWriter.Verbose(string.Format(
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: TraceWriter.Error

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+                        "Function runtime exception: {0}: {1}{2}{3}",
+                        runtimeException.ErrorRecord.InvocationInfo.InvocationName,
+                        runtimeException.Message,
+                        Environment.NewLine,
+                        runtimeException.ErrorRecord.InvocationInfo.PositionMessage));
+                }
+
+                TraceWriter.Verbose(string.Format("PowerShell Function completed (Failure, Id={0})", invocationId));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: TraceWriter.Error

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+            List<string> moduleRelativePaths = new List<string>();
+            foreach (string moduleFile in _moduleFiles)
+            {
+                string relativePath = GetRelativePath(moduleFile);
+                moduleRelativePaths.Add(relativePath);
+            }
+
+            if (moduleRelativePaths.Any())
+            {
+                TraceWriter.Info(string.Format("Loaded modules:{0}{1}", Environment.NewLine, string.Join(Environment.NewLine, moduleRelativePaths)));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Wondering if this should be Info or Verbose? Will people want to see this always? Or only opt in, if the set the level to Verbose?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+                var envSetting = Environment.GetEnvironmentVariable(pair.Key, EnvironmentVariableTarget.Process);
+                if (!string.IsNullOrEmpty(envSetting))
+                {
+                    Environment.SetEnvironmentVariable(pair.Key, null, EnvironmentVariableTarget.Process);
+                }
+            }
+        }
+
+        private async Task ProcessInputBindingsAsync(object input, string functionInstanceOutputPath, IBinderEx binder, Dictionary<string, string> bindingData, Dictionary<string, string> environmentVariables)
+        {
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Yeah, this will be a big maintenace headache. We need to refactor this into a shared base class so you can just inherit it.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+                    environmentVariables[varName] = header.Value.First();
+                }
+            }
+
+            SetProcessEnvironmentVariables(environmentVariables);
+        }
+
+        private void ReloadScript()
+        {
+            TraceWriter.Verbose(string.Format(CultureInfo.InvariantCulture, "Script for function '{0}' changed. Reloading.", Metadata.Name));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Info

File: src/WebJobs.Script/Description/Script/ScriptFunctionInvoker.cs [hunk 15]
```diff
@@ -104,7 +100,7 @@ public static bool IsSupportedScriptType(ScriptType scriptType)
                 }
             }
 
-            TraceWriter.Info(string.Format("Function started (Id={0})", invocationId));
+            TraceWriter.Info(string.Format("Script Function started (Id={0})", invocationId));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Script/ScriptFunctionInvoker.cs]: Please leave these as just "Function started", including in your PowerShell invoker. The user knows the type of their script :)

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 16]
```diff
@@ -0,0 +1,490 @@
...
+        /// Event handler for the error stream.
+        /// </summary>
+        private void ErrorDataAdded(object sender, DataAddedEventArgs e)
+        {
+            var source = (PSDataCollection<ErrorRecord>)sender;
+            var msg = GetErrorMessage(source[e.Index]);
+            TraceWriter.Error(msg);
+        }
+
+        private string GetErrorMessage(ErrorRecord errorRecord)
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Need to add lots of unit tests. Methods like this and some of your other methods are easily unit testable.

Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Also, you probably want to add some new e2e tests and samples exercising the new functionality you're lighting up (new run capabilities, modules, etc.).

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs (hunk 2 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: By the way... for the Begin/End recommendation, you should be able to just use the Task.Factory.FromAsync...

**Comment by mathewc** [comment]
Why closing this?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+                    invocationId);
+
+                Dictionary<string, string> bindingData = GetBindingData(convertedInput, binder);
+                bindingData["InvocationId"] = invocationId;
+
+                await
+                    ProcessInputBindingsAsync(convertedInput, functionInstanceOutputPath, binder, bindingData,
+                        _environmentVariables);
+
+                InitializeEnvironmentVariables(_environmentVariables, functionInstanceOutputPath, input,
```
Comment (fabiocav) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Shouldn't the environment variables here be local? We don't want to share this state between function invocations.

Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Yes, I'm expecting you to pass in this environment variables collection to the InvokePowerShellScriptCall so they are scoped to a single script execution (like they are for all other script invocations). Should probably put a test in place for this.

Comment (tohling) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: I have moved it to local.  Kindly review again.

File: WebJobs.Script.sln [hunk 18]
```diff
@@ -250,6 +250,13 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "HttpTrigger-Powershell", "H
 		sample\HttpTrigger-Powershell\run.ps1 = sample\HttpTrigger-Powershell\run.ps1
 	EndProjectSection
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "modules", "modules", "{23A9F11B-AD89-4164-BAD3-2B83B80A8173}"
+	ProjectSection(SolutionItems) = preProject
+		sample\HttpTrigger-Powershell\modules\Get-WorkerProcess.psm1 = sample\HttpTrigger-Powershell\modules\Get-WorkerProcess.psm1
```
Comment (mathewc) [inline WebJobs.Script.sln]: Please add a test to SamplesEndToEnd tests for your new sample. Take a look at the other tests - they'll enable you to verify the script log output etc.

File: sample/HttpTrigger-Powershell/function.json [hunk 19]
```diff
@@ -1,15 +1,17 @@
...
-        },
-        {
-            "type": "http",
-            "name": "res",
-            "direction": "out"
-        }
-    ]
-}
+{
+  "bindings": [
```
Comment (mathewc) [inline sample/HttpTrigger-Powershell/function.json]: Looks like the changes to this file can be reverted?

File: sample/HttpTrigger-Powershell/modules/Get-WorkerProcess.psm1 [hunk 20]
```diff
@@ -0,0 +1,5 @@
+﻿Function Get-WorkerProcess
+{
```
Comment (mathewc) [inline sample/HttpTrigger-Powershell/modules/Get-WorkerProcess.psm1]: We run these samples outside of IIS Express (using Console ScriptHost). If it's easy to modify your example so it works in that case as well it would be good.

File: sample/HttpTrigger-Powershell/run.ps1 [hunk 21]
```diff
@@ -1,10 +1,8 @@
-if ($Env:req_query_name) 
-{
```
Comment (mathewc) [inline sample/HttpTrigger-Powershell/run.ps1]: I'd like to keep this existing sample as is, since it is designed to perform the same operations as the Node/C# examples. Recommend adding a new sample for your stuff. Perhaps "HttpTrigger-Powershell-Modules" or something

File: sample/QueueTrigger-Powershell/function.json [hunk 22]
```diff
@@ -1,17 +1,19 @@
...
-            "name": "output",
-            "direction": "out",
-            "tableName": "samples",
-            "partitionKey": "samples-powershell",
-            "rowKey": "{id}"
-        }
-    ]
-}
+﻿{
+  "bindings": [
```
Comment (mathewc) [inline sample/QueueTrigger-Powershell/function.json]: Please revert this file - looks like you're not changing anything?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 23]
```diff
@@ -0,0 +1,451 @@
...
+            if (_outputBindings.Count > 0 || _inputBindings.Any())
+            {
+                Directory.CreateDirectory(functionInstanceOutputPath);
+            }
+
+            // process input bindings
+            foreach (var inputBinding in _inputBindings)
+            {
+                // $input is a reserved environment variable in PowerShell.  Using $inputdata instead.
+                string inputdata = "inputdata";
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: This will be a breaking change for existing scripts, right?

File: sample/QueueTrigger-Powershell/function.json (hunk 22 above)
Comment (davidebbo) [inline sample/QueueTrigger-Powershell/function.json]: Though I think we did decide to standardize on 2 spaces instead of 4. Should not be part of this change, but we should do a cleanup pass on all of them.

Comment (davidebbo) [inline sample/QueueTrigger-Powershell/function.json]: Also, she did add explicit authlevel and disabled.

Cool trick: when looking at CR, add `?w=1` to URL to ignore spaces. Then the change becomes clearer.

Comment (mathewc) [inline sample/QueueTrigger-Powershell/function.json]: I already did the cleanup pass on all files in this repo. They're all 4 spaces which you convinced me to move to.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 24]
```diff
@@ -0,0 +1,451 @@
...
+                TraceWriter.Info(string.Format("Function completed (Failure, Id={0})", invocationId));
+                throw;
+            }
+            finally
+            {
+                _metrics.EndEvent(startedEvent);
+            }
+        }
+
+        // private void InvokePowerShellScript()
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: nit: remove this unecessary comment?

File: sample/QueueTrigger-Powershell/function.json (hunk 22 above)
Comment (davidebbo) [inline sample/QueueTrigger-Powershell/function.json]: Oh, I thought it was the other way around :)

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 25]
```diff
@@ -0,0 +1,451 @@
...
+                    }
+                }
+
+                stackTrace = string.Join(" ", newtokens);
+            }
+
+            return stackTrace;
+        }
+
+        private string GetScript()
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Isn't there a way for us to just pass the path into the runspace, rather than reading this file ourselves?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 26]
```diff
@@ -0,0 +1,451 @@
...
+        /// </summary>
+        private void ErrorDataAdded(object sender, DataAddedEventArgs e, TraceWriter traceWriter)
+        {
+            var source = (PSDataCollection<ErrorRecord>)sender;
+            var msg = GetErrorMessage(source[e.Index]);
+            traceWriter.Error(msg);
+            TraceWriter.Error(msg);
+        }
+
+        private string GetErrorMessage(ErrorRecord errorRecord)
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Need unit tests for all of these complicated formatting functions. Make them internal and add tests.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 27]
```diff
@@ -0,0 +1,451 @@
...
+                stringBuilder.AppendLine(string.Format("{0}{1} {2} {3}",
+                    new string(PowerShellConstants.SpaceChar, PowerShellConstants.SpaceCount),
+                    PowerShellConstants.AdditionChar, PowerShellConstants.FullyQualifiedErrorIdLabel,
+                    errorRecord.FullyQualifiedErrorId));
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        private string GetStackTrace(string scriptStackTrace, string fileName)
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: Same for this function. Please make internal and add unit tests.

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 28]
```diff
@@ -0,0 +1,451 @@
...
+            string script = null;
+            if (File.Exists(_scriptFilePath))
+            {
+                script = File.ReadAllText(_scriptFilePath);
+            }
+
+            return script;
+        }
+
+        private List<string> GetModuleFilePaths()
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: This is another function that we should have a Theory based unit test for.

File: test/WebJobs.Script.Tests/TestScripts/Powershell/QueueTriggerToBlob/function.json [hunk 29]
```diff
@@ -1,15 +1,16 @@
...
-        },
-        {
-            "type": "blob",
-            "name": "output",
-            "direction": "out",
-            "path": "test-output-powershell/{id}"
-        }
-    ]
-}
+﻿{
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Powershell/QueueTriggerToBlob/function.json]: Please revert

File: sample/HttpTrigger-Powershell-Modules/modules/show-calendar.psm1 [hunk 30]
```diff
@@ -0,0 +1,147 @@
+<#
```
Comment (mathewc) [inline sample/HttpTrigger-Powershell-Modules/modules/show-calendar.psm1]: nit: can you normalize the casing of your file names? This one is lower, it's peer is upper.

File: sample/HttpTrigger-Powershell/function.json [hunk 31]
```diff
@@ -1,15 +1,15 @@
...
-        },
-        {
-            "type": "http",
-            "name": "res",
-            "direction": "out"
-        }
-    ]
-}
+{
+    "bindings": [
```
Comment (mathewc) [inline sample/HttpTrigger-Powershell/function.json]: if you're not changing this file, please revert

File: sample/HttpTrigger-Powershell/run.ps1 [hunk 32]
```diff
@@ -1,10 +1,10 @@
-if ($Env:req_query_name)
```
Comment (mathewc) [inline sample/HttpTrigger-Powershell/run.ps1]: Does this other syntax also still work (explicit reference of $Env?) I.e., are there breaking changes?

File: sample/QueueTrigger-Powershell/function.json [hunk 33]
```diff
@@ -1,17 +1,17 @@
...
-    ]
-}
+﻿{
+    "bindings": [
+        {
+            "type": "queueTrigger",
+            "direction": "in",
+            "queueName": "samples-powershell"
+        },
+        {
```
Comment (mathewc) [inline sample/QueueTrigger-Powershell/function.json]: revert this file

File: sample/HttpTrigger-Powershell/run.ps1 (hunk 32 above)
Comment (tohling) [inline sample/HttpTrigger-Powershell/run.ps1]: No, $Env will not work any longer.  The environment variables are now passed in as variables into the runspace that the PowerShell instance runs in.  As such, all environment variables will be treated as variables, e.g. $Env:req_query_name => $req_query_name, with the exception of $Env:input, because $input is a reserved variable in PowerShell. As such, the default input parameter will be $inputdata.

In the event that the user uses "input", then an exception will be thrown. See,

src/WebJobs.Script/Description/FunctionDescriptorProvider.cs

and accompanying tests,

test/WebJobs.Script.Tests/TestScripts/PowershellFunctionGenerationTests.cs
- GenerateQueueTriggerFunction
- GenerateQueueTriggerFunction_WithInvalidInputName_Fails

File: sample/QueueTrigger-Powershell/queueTrigger.ps1 [hunk 34]
```diff
@@ -1,8 +1,4 @@
-$in = Get-Content $Env:input
-
-[Console]::WriteLine("Powershell script processed queue message '$in'")
-
-$output = $Env:output
-$json = $in | ConvertFrom-Json
-$entity = [string]::Format('{{ "Status": 0, "Title": "Powershell Table Entity for message {0}" }}', $json.id)
+$in = Get-Content $inputdata
```
Comment (mathewc) [inline sample/QueueTrigger-Powershell/queueTrigger.ps1]: If you used $Env:input here, can we avoid the breaking change of renaming input to inputdata, and retain consistency across the languages?

File: sample/HttpTrigger-Powershell/run.ps1 (hunk 32 above)
Comment (mathewc) [inline sample/HttpTrigger-Powershell/run.ps1]: I see, that clarifies, thanks.

File: src/WebJobs.Script/Description/FunctionDescriptorProvider.cs [hunk 35]
```diff
@@ -18,6 +18,7 @@ namespace Microsoft.Azure.WebJobs.Script.Description
     public abstract class FunctionDescriptorProvider
     {
         internal const string DefaultInputParameterName = "input";
+        internal const string DefaultPowershellInputParameterName = "inputdata";
```
Comment (mathewc) [inline src/WebJobs.Script/Description/FunctionDescriptorProvider.cs]: I'm not sure, but don't the design guidelines for PowerShell use camel case for variable names? https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx. If so, this should be **inputData**

File: sample/HttpTrigger-Powershell/run.ps1 (hunk 32 above)
Comment (mathewc) [inline sample/HttpTrigger-Powershell/run.ps1]: Will you also be handling the updates in the templates repo? There are several PS templates I believe. https://github.com/Azure/azure-webjobs-sdk-templates/blob/dev/Templates/QueueTrigger-Powershell/run.ps1

File: src/WebJobs.Script/Description/FunctionDescriptorProvider.cs [hunk 36]
```diff
@@ -53,9 +54,17 @@ public virtual bool TryCreate(FunctionMetadata functionMetadata, out FunctionDes
                     }
                     else
                     {
-                        binding.Name = DefaultInputParameterName;
+                        ScriptType scriptType = functionMetadata.ScriptType;
+                        binding.Name = scriptType == ScriptType.Powershell ? DefaultPowershellInputParameterName : DefaultInputParameterName;
                     }
                 }
+                else if (functionMetadata.ScriptType == ScriptType.Powershell &&
+                         string.Equals(binding.Name, DefaultInputParameterName, StringComparison.OrdinalIgnoreCase))
```
Comment (mathewc) [inline src/WebJobs.Script/Description/FunctionDescriptorProvider.cs]: use "input" literal directly here, rather than DefaultInputParameterName since we're explicitly guarding against that. If we change DefaultInputParameterName  in the future, you want this logic to remain unchanged.

File: src/WebJobs.Script/Description/FunctionDescriptorProvider.cs [hunk 37]
```diff
@@ -53,9 +54,17 @@ public virtual bool TryCreate(FunctionMetadata functionMetadata, out FunctionDes
...
+                {
+                    throw new ArgumentException(
+                        string.Format("Input binding name '{0}' not allowed.  Consider renaming.",
```
Comment (mathewc) [inline src/WebJobs.Script/Description/FunctionDescriptorProvider.cs]: nit: "Input binding name 'input' is not allowed."

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 38]
```diff
@@ -0,0 +1,451 @@
...
+
+                await
+                    ProcessOutputBindingsAsync(functionInstanceOutputPath, _outputBindings, input, binder, bindingData);
+
+                if (errors.Any())
+                {
+                    ErrorRecord error = errors.FirstOrDefault();
+                    if (error != null)
+                    {
+                        throw new RuntimeException("Powershell script error", error.Exception, error);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: nit: PowerShell?

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 39]
```diff
@@ -0,0 +1,451 @@
...
+                }
+                else
+                {
+                    TraceWriter.Info(string.Format("Function completed (Success, Id={0})", invocationId));
+                }
+            }
+            catch (Exception)
+            {
+                startedEvent.Success = false;
+                TraceWriter.Info(string.Format("Function completed (Failure, Id={0})", invocationId));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: TraceWriter.Error (I mentioned this before)

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs [hunk 40]
```diff
@@ -0,0 +1,451 @@
...
+                modulePaths.AddRange(Directory.GetFiles(moduleDirectory,
+                    PowerShellConstants.ModulesBinaryFileExtensionPattern));
+                modulePaths.AddRange(Directory.GetFiles(moduleDirectory,
+                    PowerShellConstants.ModulesScriptFileExtensionPattern));
+            }
+
+            return modulePaths;
+        }
+
+        private void InitializeEnvironmentVariables(Dictionary<string, string> environmentVariables, string functionInstanceOutputPath, object input, ExecutionContext context)
```
Comment (mathewc) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: I believe this function is a complete unmodified duplicate of ScriptFunctionInvoker.InitializeEnvironmentVariables. Please add a new ScriptFunctionInvoker base class (that itself derives from FunctionInvoker), and add this as a protected method. That way your code and ScriptFunctionInvoker can call it.

File: test/WebJobs.Script.Tests/PowershellEndToEndTests.cs [hunk 41]
```diff
@@ -1,6 +1,18 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/PowershellEndToEndTests.cs]: nit: fix PowerShell casing on new files

File: test/WebJobs.Script.Tests/TestScripts/Powershell/BlobTriggerToBlob-Powershell/function.json [hunk 42]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Powershell/BlobTriggerToBlob-Powershell/function.json]: remove the "-Powershell" suffix on this folder, since it is already scoped to PowerShell.

File: test/WebJobs.Script.Tests/TestScripts/Powershell/HttpTrigger-Powershell/run.ps1 [hunk 43]
```diff
@@ -0,0 +1,21 @@
...
+}
+catch
+{
+	$requestBody = Get-Content $req -Raw
+}
+ 
+if ($req_query_name) 
+{
+    $name = $req_query_name;
+	$message = "Hello $name";
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Powershell/HttpTrigger-Powershell/run.ps1]: nit: fix indent

File: src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs (hunk 40 above)
Comment (tohling) [inline src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs]: I don't see a FunctionInvoker class.  Did you mean IFunctionInvoker or FunctionInvokerBase?

File: test/WebJobs.Script.Tests/TestScripts/Powershell/HttpTrigger-PowershellWithError/run.ps1 [hunk 44]
```diff
@@ -0,0 +1,14 @@
+$dateObject = Get-DateToday;
+$dateMessage = "Today's date is $dateObject";
+if ($req_query_name) 
+{
+    $name = $req_query_name;
+	$message = "Hello $name.  Today's date is $dateMessage";
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Powershell/HttpTrigger-PowershellWithError/run.ps1]: fix indenting in this file

**Comment by mathewc** [comment]
Merged :)

**Comment by fabiocav** [comment]
🎉
